File: src/main/java/ru/org/linux/admin/ipmanage/SameIpService.java
/*
 * Copyright 1998-2011 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.admin.ipmanage;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import ru.org.linux.message.MessageNotFoundException;
import ru.org.linux.user.User;
import ru.org.linux.user.UserDao;
import ru.org.linux.user.UserNotFoundException;

import java.util.ArrayList;
import java.util.List;

@Service
public class SameIpService {

  @Autowired
  SameIpDao sameIpDao;

  @Autowired
  private IPBlockDao ipBlockDao;

  @Autowired
  private UserDao userDao;


  /**
   * Получить список пользователей, заходивших с указанного IP-адреса.
   *
   * @param ipAddress IP-адрес
   * @param uaId      - идентификатор строки UserAgent
   * @return Список пользователей
   */
  public List<SameIp.UserItem> getUsers(String ipAddress, final int uaId) {
    List<SameIpDto.UserItem> userItemsDto = sameIpDao.getUsers(ipAddress);
    List<SameIp.UserItem> userItems = new ArrayList<SameIp.UserItem>();

    for (SameIpDto.UserItem userItemDto : userItemsDto) {
      userItems.add(new SameIp.UserItem(userItemDto, uaId));
    }
    return userItems;
  }

  /**
   * Получить список тем или комментариев, написанных с указанного IP-адреса.
   *
   * @param ipAddress IP-адрес
   * @param isComment true если необходимо получить комментарии; false если необходимо получить темы
   * @return список тем или комментариев
   */
  public List<SameIp.TopicItem> getForumMessages(String ipAddress, boolean isComment) {
    List<SameIpDto.TopicItem> topicItemsDto;

    topicItemsDto = isComment
      ? sameIpDao.getComments(ipAddress)
      : sameIpDao.getTopics(ipAddress);

    List<SameIp.TopicItem> topicItems = new ArrayList<SameIp.TopicItem>();

    for (SameIpDto.TopicItem topicItemDto : topicItemsDto) {
      topicItems.add(new SameIp.TopicItem(topicItemDto, isComment));
    }
    return topicItems;
  }

  /**
   * Получение информации о блокировке указанного адреса.
   *
   * @param ipAddress IP-адрес
   * @return объект, содержащий информацию блокировке IP-адреса
   */
  public SameIp.BlockInfo getBlockInfo(String ipAddress) {
    SameIp.BlockInfo blockInfo = new SameIp.BlockInfo();

    IPBlockInfo ipBlockInfo = ipBlockDao.getBlockInfo(ipAddress);
    blockInfo.setBlocked(ipBlockInfo != null);

    if (ipBlockInfo != null) {
      blockInfo.setBanDate(ipBlockInfo.getBanDate());
      blockInfo.setOriginalDate(ipBlockInfo.getOriginalDate());
      blockInfo.setReason(ipBlockInfo.getReason());
      blockInfo.setBlockExpired(!ipBlockInfo.isBlocked());
      User moderator;
      try {
        moderator = userDao.getUserCached(ipBlockInfo.getModerator());
        blockInfo.setModeratorNick(moderator.getNick());
      } catch (UserNotFoundException ignored) {
        blockInfo.setModeratorNick("Модератор не найден");
      }
    }
    return blockInfo;
  }

  /**
   * получить информацию об IP-адресе и об идентификаторе строки  userAgent по идентификатору сообщения.
   *
   * @param msgId идентификатор сообщения
   * @return объект, содержащий информацию об IP-адресе и об идентификаторе строки userAgent
   * @throws MessageNotFoundException если сообщение не найдено.
   */
  public SameIp.IpInfo getIpInfo(Integer msgId)
    throws MessageNotFoundException {
    return sameIpDao.getIpInfo(msgId);
  }
}

File: src/main/scala/ru/org/linux/sameip/SameIpService.scala
/*
 * Copyright 1998-2024 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.sameip

import org.springframework.stereotype.Service
import ru.org.linux.markup.MessageTextService
import ru.org.linux.spring.dao.MsgbaseDao
import ru.org.linux.user.UserService

@Service
class SameIpService(userService: UserService, msgbaseDao: MsgbaseDao, textService: MessageTextService,
                    sameIpDao: SameIpDao) {
  def getPosts(ip: Option[String], userAgent: Option[Int], score: Option[Int], limit: Int): collection.Seq[PreparedPostListItem] = {
    prepareCommentList(sameIpDao.getComments(ip, userAgent, score, limit))
  }

  private def prepareCommentList(items: collection.Seq[PostListItem]): collection.Seq[PreparedPostListItem] = {
    val users = userService.getUsersCachedMap(items.map(_.authorId))
    val texts = msgbaseDao.getMessageText(items.map(i => i.commentId.getOrElse(i.topicId)))

    items.map { item =>
      val plainText = textService.extractPlainText(texts(item.commentId.getOrElse(item.topicId)))
      val textPreview = MessageTextService.trimPlainText(plainText, 250, encodeHtml = false)

      PreparedPostListItem(
        link = item.commentId.map { commentId =>
          s"jump-message.jsp?msgid=${item.topicId}&amp;cid=$commentId"
        }.getOrElse(s"jump-message.jsp?msgid=${item.topicId}"),
        groupTitle = item.groupTitle,
        author = users(item.authorId),
        title = item.title,
        deleted = item.deleted,
        textPreview = textPreview,
        reason = item.reason,
        postdate = item.postdate,
        comment = item.commentId.isDefined)
    }
  }
}
